Documentation/RepoApi.md

```diff
@@ -0,0 +1,137 @@
+One slight variation here is the usage of `-p` instead of `/p`.  MSBuild supports both forms, but source-build chooses `-p` because
+it meets [POSIX recommendations](https://www.gnu.org/software/libc/manual/html_node/Argument-Syntax.html) for command line parameters
+and avoids any confusion with file paths on Unix machines.
+
+In order to avoid conflicts between existing MSBuild properties used by the repos and the properties passed in by source-build,
+source-build will append a prefix to every property it passes in: `DotNet`.
+
+#### Detecting build from source
+
+The first parameter used by source-build will be a value indicating whether the build is being performed in "source only" mode. This
+means repos cannot use tools/packages/references that haven't already been built by the current build.
+
+`-p:DotNetBuildFromSource=true`
+
+Some repos build more projects than is necessary to build the .NET Core stack. For example, the `dotnet/roslyn` repo builds
+the C# compiler, which is needed, but it also builds Visual Studio extensions, which aren't needed. These VS extensions
+cannot be built without using an external dependency on tooling that isn't available when building from source.
+
+#### Available tools
+
+In order for repos to build, they need tools in order to do the build (ex. MSBuild). These tools are not generally available
+on all supported operating systems. As such, many repos today have the ability to pull these tools from the internet during the
+build.
+
+However, there are a couple problems with each repo downloading its own toolset.
+
+1. Different versions can cause inconsistencies
+1. It takes a non-trivial amount of time to download the toolset, if every repo does it during the complete .NET Core build,
+it will add a lot of overhead to the build.
+
+To solve this, source-build will have a set of common tools available that it will pass into each repos build.
+
+First, a [.NET Core SDK](https://www.microsoft.com/net/download/core#/sdk) will be available for repos to be able to invoke
+MSBuild, NuGet, etc. The path to the folder containing the `dotnet` executable will be available using the
+
+`-p:DotNetToolDir`
+
+property. This can be used to call `dotnet restore`, `dotnet build`, `dotnet publish`, `dotnet msbuild`, or any other CLI
+verb on the repo's projects.
+
+Second, a common set of tools used by the `dotnet` org repos is [BuildTools](https://github.com/dotnet/buildtools). This will
+be available using the
+
+`-p:DotNetBuildToolsDir`
+
+property. This can be used to invoke any current or future tool that is part of BuildTools. It is not required that each repo
+use BuildTools, but if the repo wants to, it can get the tools from this property.
+
+### Example
+
+Here is an example of the list of calls source-build will make on the `dotnet/corefx` repo:
+
+```
+/path/to/corefx/build.sh -t:Restore -p:DotNetBuildFromSource=true -p:DotNetToolDir=/path/to/dotnet -p:DotNetBuildToolsDir=/path/to/buildtools
```

Comment:
For the example of calling Restore/Build/Publish do we expect that we will call each for each repo? If so is there any great need to have them be different? The only reason I think having them be different is interesting is if we are going to call the verb on each repo together before calling the next verb on all the repos. However I don’t think we can do that in a build-from-source scenario because we have to wait to call Restore in corefx until the coreclr repo is build for example.

Reply:
The idea of having a `Restore` step that was separate from `Build` was that if `Restore` failed, it could be retried.  Maybe it failed for a network issue.

But as stated above, I'm going to remove Restore and Publish, since these were only supposed to be example commands.